File: devtools/sdk/python/cloudmock/connection.py
"""TCP JSON-line client for the CloudMock devtools source server.

Auto-reconnects every 5 seconds, buffers up to 100 messages when disconnected,
and operates entirely on a background daemon thread so it never blocks the caller.
"""

import json
import socket
import threading
import time
from collections import deque
from typing import Any, Dict, Optional

MAX_BUFFER_SIZE = 100
RECONNECT_DELAY = 5.0
CONNECT_TIMEOUT = 3.0


class Connection:
    """TCP JSON-line connection to the devtools source server."""

    def __init__(self, host: str, port: int, app_name: str) -> None:
        self.host = host
        self.port = port
        self.app_name = app_name

        self._sock: Optional[socket.socket] = None
        self._connected = False
        self._closed = False
        self._lock = threading.Lock()
        self._buffer: deque[str] = deque(maxlen=MAX_BUFFER_SIZE)

        # Start the connection thread as a daemon so it doesn't prevent exit
        self._thread = threading.Thread(target=self._connect_loop, daemon=True)
        self._thread.start()

    def _connect_loop(self) -> None:
        """Background loop that maintains the TCP connection."""
        # Attempt first connection immediately
        self._try_connect()

        while True:
            with self._lock:
                if self._closed:
                    return

            time.sleep(RECONNECT_DELAY)

            with self._lock:
                if self._closed:
                    return
                needs_connect = not self._connected

            if needs_connect:
                self._try_connect()

    def _try_connect(self) -> None:
        """Attempt a single TCP connection to the devtools server."""
        with self._lock:
            if self._closed:
                return

        try:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(CONNECT_TIMEOUT)
            sock.connect((self.host, self.port))
            sock.settimeout(None)
        except OSError:
            return

        with self._lock:
            if self._closed:
                sock.close()
                return

            self._sock = sock
            self._connected = True

            # Flush buffered messages
            failed: list[str] = []
            while self._buffer:
                msg = self._buffer.popleft()
                if not self._write_locked(msg):
                    failed.append(msg)
                    break
            for msg in failed:
                self._buffer.appendleft(msg)

        # Start a reader thread to detect disconnection
        reader = threading.Thread(target=self._read_until_close, args=(sock,), daemon=True)
        reader.start()

    def _read_until_close(self, sock: socket.socket) -> None:
        """Read from the socket until it closes (we're send-only)."""
        try:
            while True:
                data = sock.recv(256)
                if not data:
                    break
        except OSError:
            pass
        finally:
            with self._lock:
                if self._sock is sock:
                    self._connected = False
                    self._sock = None
                    try:
                        sock.close()
                    except OSError:
                        pass

    def send(self, event_type: str, data: Dict[str, Any]) -> None:
        """Send an event to the devtools server. Non-blocking."""
        event = {
            "type": event_type,
            "data": data,
            "source": self.app_name,
            "runtime": "python",
            "timestamp": int(time.time() * 1000),
        }

        try:
            msg = json.dumps(event, default=str) + "\n"
        except (TypeError, ValueError):
            return

        with self._lock:
            if self._closed:
                return

            if self._connected and self._sock is not None:
                if self._write_locked(msg):
                    return
                # Write failed; mark disconnected
                self._connected = False
                if self._sock is not None:
                    try:
                        self._sock.close()
                    except OSError:
                        pass
                    self._sock = None

            # Buffer (deque with maxlen handles overflow automatically)
            self._buffer.append(msg)

    def _write_locked(self, msg: str) -> bool:
        """Write a message to the socket. Must be called with _lock held.

        Returns True on success, False on failure.
        """
        if self._sock is None:
            return False
        try:
            self._sock.sendall(msg.encode("utf-8"))
            return True
        except OSError:
            return False

    def close(self) -> None:
        """Close the connection and stop reconnection attempts."""
        with self._lock:
            if self._closed:
                return
            self._closed = True
            self._connected = False
            if self._sock is not None:
                try:
                    self._sock.close()
                except OSError:
                    pass
                self._sock = None
            self._buffer.clear()

File: devtools/sdk/kotlin/src/main/kotlin/io/cloudmock/sdk/Connection.kt
package io.cloudmock.sdk

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import java.io.BufferedWriter
import java.io.OutputStreamWriter
import java.net.Socket
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean

/**
 * TCP JSON-line client that connects to the devtools source server.
 * Runs on a background daemon thread to avoid blocking the main thread.
 * Auto-reconnects every 5 seconds and buffers up to 100 messages when disconnected.
 */
internal class Connection(
    private val host: String,
    private val port: Int,
    val appName: String
) {
    private var socket: Socket? = null
    private var writer: BufferedWriter? = null
    private val connected = AtomicBoolean(false)
    private val closed = AtomicBoolean(false)
    private val buffer = ConcurrentLinkedQueue<String>()
    private val maxBufferSize = 100

    private val json = Json { encodeDefaults = true }

    init {
        startConnectionThread()
    }

    private fun startConnectionThread() {
        val thread = Thread {
            while (!closed.get()) {
                try {
                    connect()
                } catch (_: Exception) {
                    // Connection failed; wait and retry
                }

                if (!closed.get()) {
                    try {
                        Thread.sleep(5000)
                    } catch (_: InterruptedException) {
                        break
                    }
                }
            }
        }
        thread.isDaemon = true
        thread.name = "cloudmock-connection"
        thread.start()
    }

    private fun connect() {
        if (closed.get()) return

        val sock = Socket(host, port)
        socket = sock
        writer = BufferedWriter(OutputStreamWriter(sock.getOutputStream(), Charsets.UTF_8))
        connected.set(true)

        // Flush buffered messages
        flushBuffer()

        // Keep connection alive by reading (blocks until disconnect)
        try {
            val input = sock.getInputStream()
            val buf = ByteArray(256)
            while (!closed.get() && !sock.isClosed) {
                val bytesRead = input.read(buf)
                if (bytesRead == -1) break
                // Discard any server data; we're send-only
            }
        } catch (_: Exception) {
            // Connection lost
        } finally {
            connected.set(false)
            closeSocket()
        }
    }

    /**
     * Send an event to the devtools server.
     * Thread-safe; buffers messages if disconnected.
     */
    fun send(event: SourceEvent) {
        val line = serializeEvent(event) ?: return
        val message = "$line\n"

        if (connected.get()) {
            try {
                synchronized(this) {
                    writer?.write(message)
                    writer?.flush()
                }
                return
            } catch (_: Exception) {
                // Write failed; buffer it
            }
        }

        // Buffer up to maxBufferSize messages
        if (buffer.size < maxBufferSize) {
            buffer.offer(message)
        }
    }

    private fun flushBuffer() {
        val w = writer ?: return
        try {
            synchronized(this) {
                while (true) {
                    val msg = buffer.poll() ?: break
                    w.write(msg)
                }
                w.flush()
            }
        } catch (_: Exception) {
            // Flush failed
        }
    }

    private fun serializeEvent(event: SourceEvent): String? {
        return try {
            val obj = buildJsonObject {
                put("type", event.type)
                put("source", event.source)
                put("runtime", event.runtime)
                put("timestamp", event.timestamp)
                put("data", serializeMap(event.data))
            }
            json.encodeToString(obj)
        } catch (_: Exception) {
            null
        }
    }

    private fun serializeMap(map: Map<String, Any?>): JsonElement {
        return buildJsonObject {
            for ((key, value) in map) {
                put(key, toJsonElement(value))
            }
        }
    }

    private fun toJsonElement(value: Any?): JsonElement {
        return when (value) {
            null -> JsonNull
            is String -> JsonPrimitive(value)
            is Number -> JsonPrimitive(value)
            is Boolean -> JsonPrimitive(value)
            is Map<*, *> -> buildJsonObject {
                @Suppress("UNCHECKED_CAST")
                for ((k, v) in value as Map<String, Any?>) {
                    put(k, toJsonElement(v))
                }
            }
            is List<*> -> buildJsonArray {
                for (item in value) {
                    add(toJsonElement(item))
                }
            }
            else -> JsonPrimitive(value.toString())
        }
    }

    private fun closeSocket() {
        try {
            writer?.close()
        } catch (_: Exception) {}
        try {
            socket?.close()
        } catch (_: Exception) {}
        writer = null
        socket = null
    }

    /**
     * Close the connection and stop reconnection attempts.
     */
    fun close() {
        closed.set(true)
        connected.set(false)
        closeSocket()
        buffer.clear()
    }
}

/**
 * Event sent to the devtools source server.
 */
internal data class SourceEvent(
    val type: String,
    val data: Map<String, Any?>,
    val source: String,
    val runtime: String,
    val timestamp: Long = System.currentTimeMillis()
)
